docs/SIL.rst

```diff
@@ -2043,6 +2043,148 @@ parts::
+not be used after the underlying object has been deallocated. To see this,
+consider the following C++ example::
+
+    int unfortunateFunction() {
+      int *unsafeInteriorPointer = nullptr;
+      {
+        std::vector<int> vector;
+        vector.push_back(5);
+        unsafeInteriorPointer = vector.data();
+        printf("%d\n", *unsafeInteriorPointer); // Prints "5".
+      } // vector deallocated here
+      return *unsafeInteriorPointer; // Kaboom
+    }
+
+In words, C++ allows for us to get the interior pointer into the vector, but
+then lets us do whatever we want with the pointer, including use it after the
+underlying memory has been invalidated.
+
+From a user’s perspective, interior pointers are really useful since one can use
+it to pass data to other APIs that are only expecting a pointer and also since
+one can use it to sometimes get better performance. But from a language designer
+perspective, this sort of API verbotten and leads to bugs, crashes, and security
+vulnerabilities. That being said, clearly users have a need for such
+functionality, so we, as language designers, should figure out manners to
+express these sorts of patterns in our various languages in a safe way that
+prevents user’s from foot-gunning themselves. In SIL, we have solved this
+problem via the direct modeling of interior pointer instructions as a high level
+concept in our IR.
+
+Safe Interior Pointers in SIL
+`````````````````````````````
+
+In contrast to LLVM-IR, SIL provides mechanisms that language designers can use
+to express concepts like the above in a manner that allows the language to
+define away compiler generated unsafe interior pointer usage using "Safe
+Interior Pointers". This is implemented in SIL by:
+
+1. Classifying a set of instructions as being "interior pointer" instructions.
+2. Enforcing in the SILVerifier that all "interior pointer" instructions can
+   only have operands with `Guaranteed`_ ownership.
+3. Enforcing in the SILVerifier that any transitive address use of the interior
+   pointer to be a liveness requirement of the "interior pointer"'s
+   operand.
+
+This ensures statically that any uses of the address that are not escaped to be
+an object type (e.x.: `address_to_pointer`_) are within the guaranteed pointers
+scope where the guaranteed value is statically known to be live. As a result, in
+SIL it is impossible to express such a bug in compiler generated code. As an
+example, consider the following unsafe interior pointer SIL::
+
+    // Today SIL restricts interior pointer instructions to only have operands
+    // with guaranteed ownership. The author thinks it could be implemented
+    // sometime in the future.
+    %1 = begin_borrow %0 : $Klass
+
+    // %2 is an interior pointer into %1. Since %2 is an address, it's uses are
+    // not treated as uses of the underlying borrowed object %1 in the ownership
+    // system. This is because at the ownership level objects with none
```

Comment:
"with _.none_ ownership"

Reply:
I am going to `None`_ so it jumps back, but good catch.

Reply:
Ah. Actually I didn't see this is in line SIL so I will do None. We shouldn't use .none or it sounds like we are talking about optionals.

Reply:
OK, sounds good. There's another place that needs to be changed too, then.